### PR TITLE
Use simplified vector implementations with resolution translation

### PR DESCRIPTION
**Pull request by ZehMatt**
Instead of generating more and more 'vector' types we can use two generic implementations which implements 2 and 3 dimensional vectors. All the location/title_pos/map_pos/map_pos3 are vectors with different resolution/types but nonetheless still vectors.

Since I don't play the game much I can not tell if I actually broke something, but from what I can tell it seems fine. This also caught a bug with the volume calculation, I'll remark it in the comments.

### CODE REVIEW

File: src/OpenLoco/Audio/Audio.cpp [hunk 1]
```diff
@@ -613,7 +613,7 @@ namespace OpenLoco::Audio
                     return;
                 }
 
-                volume += calculateVolumeFromViewport(id, { loc.x, loc.y }, *viewport);
+                volume += calculateVolumeFromViewport(id, { loc.x, loc.y, 0 }, *viewport);
```
Comment (ZehMatt) [inline src/OpenLoco/Audio/Audio.cpp:616]: This did previously always pass ```z = 0``` because the implementation allowed implicit construction of z which is always zero. The new implementation does not allow this and it should never allow it, things like this can happen.

File: src/OpenLoco/Math/Vector.hpp [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+            case 3:
+                res.x = -vec.y;
+                res.y = vec.x;
+                break;
+        }
+        return res;
+    }
+
+    template<typename T, T TResolution>
+    static constexpr auto Distance(const TVector2<T, TResolution>& lhs, const TVector2<T, TResolution>& rhs)
```
Comment (duncanspumpkin) [inline src/OpenLoco/Math/Vector.hpp:260]: ManhattanDistance

We don't want to get this confused with computing the magnitude between two points which there is also equations for.

Comment (ZehMatt) [reply src/OpenLoco/Math/Vector.hpp:260]: Good point

File: src/OpenLoco/Paint/PaintEntity.h [hunk 3]
```diff
@@ -1,11 +1,7 @@
 #pragma once
+#include "../Map/Map.hpp"
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Paint/PaintEntity.h:2]: Do we really need to include this header in this small one?

File: src/OpenLoco/Paint/Paint.h [hunk 4]
```diff
@@ -1,12 +1,11 @@
 #pragma once
 #include "../Interop/Interop.hpp"
+#include "../Map/Map.hpp"
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Paint/Paint.h:3]: Do we really need to include this header in this small one?

File: src/OpenLoco/Audio/Audio.cpp [hunk 5]
```diff
@@ -613,7 +613,7 @@ namespace OpenLoco::Audio
                     return;
                 }
 
-                volume += calculateVolumeFromViewport(id, { loc.x, loc.y }, *viewport);
+                volume += calculateVolumeFromViewport(id, { loc.x, loc.y, loc.z }, *viewport);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Audio/Audio.cpp:616]: This may be worth spinning off to its own PR.

**Comment by AaronVanGeffen** [comment]
Fwiw, I think we should hold off on merging this until after the release we plan to do this week.

File: src/OpenLoco/Paint/PaintEntity.h (hunk 3 above)
Comment (ZehMatt) [reply src/OpenLoco/Paint/PaintEntity.h:2]: I either include Map.hpp which already defines the types or I need to include Vector.hpp and define it again. Forward declaring the vector template is a bit ugly honestly.

File: src/OpenLoco/Math/Vector.hpp [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        }
+
+        constexpr TVector2& operator/=(const T rhs)
+        {
+            x /= rhs;
+            y /= rhs;
+            return *this;
+        }
+
+        constexpr bool operator>=(const TVector2& rhs) const
```
Comment (duncanspumpkin) [inline src/OpenLoco/Math/Vector.hpp:79]: I still don't see the point of these comparisons but I'm not overly fussed. I guess it might be useful for some ordering container.

**Review by duncanspumpkin — APPROVED** [review]

**Comment by AaronVanGeffen** [comment]
I was as little concerned about compilation times, so I timed how long it takes to cleanly compile this PR and its parent commit. (Cleaned my working directory and ccache before timing.)

This PR:
```
$ time ninja
real    1m23.202s
user    4m17.552s
sys     0m20.177s
```

5c90cdb5 (its parent):
```
real    1m16.313s
user    4m17.511s
sys     0m19.463s
```

I think we can spare the added six seconds.

File: src/OpenLoco/Audio/Audio.cpp (hunk 5 above)
Comment (AaronVanGeffen) [inline src/OpenLoco/Audio/Audio.cpp:616]: We should probably add this change to the changelog at least.

File: src/OpenLoco/Math/Vector.hpp (hunk 6 above)
Comment (ZehMatt) [reply src/OpenLoco/Math/Vector.hpp]: Removed the unused operators

File: src/OpenLoco/Audio/Audio.cpp (hunk 5 above)
Comment (ZehMatt) [reply src/OpenLoco/Audio/Audio.cpp:616]: Added the entry
